pocs/serial_handlers/protocol_arduinosimulator.py

```diff
@@ -96,47 +107,47 @@ def run(self):
-        while not self.relay_queue.empty():
-            b = self.relay_queue.get_nowait()
-            assert isinstance(b, (bytes, bytearray))
-            self.pending_relay_bytes.extend(b)
-        self.handle_pending_relay_bytes()
-
     def handle_pending_relay_bytes(self):
-        """Process complete relay commands. NOT YET IMPLEMENTED."""
-        if len(self.pending_relay_bytes):
-            self.logger.error('ArduinoSimulator.handle_pending_relay_bytes NOT YET IMPLEMENTED')
-            self.logger.error('Dropping bytes: {!r}', self.pending_relay_bytes)
-            self.pending_relay_bytes.clear()
-        pass
+        """Process complete relay commands."""
+        newline = b'\n'
+        while True:
+            index = self.pending_relay_bytes.find(newline)
+            if index < 0:
+                break
+            line = str(self.pending_relay_bytes[0:index], 'ascii')
+            self.logger.info(f'Received command: {line}')
+            del self.pending_relay_bytes[0:index + 1]
+            self.command_lines.append(line)
+        if self.pending_relay_bytes:
+            self.logger.info(f'Accumulated {len(self.pending_relay_bytes)} bytes.')
 
     def read_relay_queue_until(self, next_time):
-        """Wait for a little bit for relay input, not more than the time until the next action."""
-        now = datetime.datetime.now()
-        if now >= next_time:
-            # Already reached the time for the next main loop event,
-            # so return to repeat the main loop.
-            return
-        remaining = (next_time - now).total_seconds()
-        assert remaining > 0
-        self.logger.info('ArduinoSimulator.read_relay_queue_until remaining={}', remaining)
-        try:
-            b = self.relay_queue.get(block=True, timeout=remaining)
-        except queue.Empty:
-            return
-        if b:
-            assert isinstance(b, (bytes, bytearray))
-            self.pending_relay_bytes.extend(b)
-            self.handle_pending_relay_bytes()
+        """Read and process relay queue bytes until time for the next action."""
+        while True:
+            now = datetime.datetime.now()
+            if now >= next_time:
+                # Already reached the time for the next main loop event,
+                # so return to repeat the main loop.
+                return
+            remaining = (next_time - now).total_seconds()
+            assert remaining > 0
+            self.logger.info('ArduinoSimulator.read_relay_queue_until remaining={}', remaining)
+            try:
+                b = self.relay_queue.get(block=True, timeout=remaining)
+                assert isinstance(b, (bytes, bytearray))
+                self.pending_relay_bytes.extend(b)
+                self.handle_pending_relay_bytes()
+                # Fake a baud rate for reading by waiting based on the
+                # number of bytes we just read.
+                time.sleep(1.0 / 1000 * len(b))
```

Comment:
Can `b` ever be zero?

Reply:
In practice, no it can't. Fortunately, if that happened, the length of the sleep would be zero, which is just fine.